### ISSUE TITLE
What are the channels and how to configure them?

### ISSUE THREAD

**Issue by Maddmax76** [mhaack/mqtt-433mhz-gateway-homie#3]
Hi, what do you mean with channels? They look like the RF contact codes, is that it?
What should I do to configure contacts to open doors, pIRs and remote controls with 4 buttons?

does the name "o-N" indicate an array?

Could you give me some guidance on how to configure my sensors?

thanks a lot

**Comment by mhaack** [comment]
See the example in readme. Channels are create the link between MQTT messages and RF codes, for example for a for button remote: `b-1:[139400];b-2:[139707];b-3:[1398097,1398100];b-4:[139503]` 

This is a list with the following pattern: ` s1:[s2]`. Where: 
* s1 = string used for the MQTT topic, you can use any string you like for example `b-1` for button one
* s2 = the RF code send or received, this can be a list for example of a cheap remote control sends not one unique code
 
A button press on button 1 (with sample above) will publish the RC code value on `homie/<device-name>/433toMQTT/channel-b-1`.

HTH

**Comment by Maddmax76** [comment]
Hi Mark, thanks for your reply.
What I miss is physically like telling him that the object that transmits
only the radio-frequency code "524745" is a read-only code, which is the
entrance door and that when that code arrives, it means that the door is
open.
Is this a gateway thing or should I implement it on the HA system?
If it does the gateway, does it in OTA or do I need to send the
configuration string via a clinetMQTT?

Specifically of my door opening sensor, the string should be:
  Door-Entry: [524745]
and this should publish on
  homie / NodeMCU / 433toMQTT / Door-Entry /

A remote control with 4 buttons configure it with:
Remote Control: [124589,124578,124536,124568]
and this should publish on
homie / NodeMCU / 433toMQTT / Remote control /?

Thanks and sorry if I'm a bit slow to understand :-)

**Comment by Maddmax76** [comment]
Excuse me, a configuration like the following is correct?

    "settings": {
        "RemoteControl": "Button-Engage:[139400];Button-Disengage:[139707];Button.Perimetric:[1398097];Button-Panic:[139803];“,
        "Door-Entry":"Active:[524745]",
         "Door-Kitchen":"Active:[524745]",
    }

**Comment by mhaack** [comment]
Please have a look at the readme for the config example. Also see Home doc (https://homieiot.github.io/homie-esp8266/docs/2.0.0/configuration/json-configuration-file/)

It should look like ```
{
    "settings": {
        "channels": "Button-Engage:[139400];Button-Disengage:[139707];Button-Perimetric:[1398097];Button-Panic:[139803];Door-Entry":[524745];Door-Kitchen:[524745];“
    }
}```

Make sure to not put any line breaks or space into the config string. Also I recommend to use shorter names like `b-e` etc. since there is a limitation on the size of the entire `settings` object.

**Comment by Maddmax76** [comment]
perfect, thank you very much, now it's much clearer.
In practice, the remote control is seen as 4 different push-button devices.

The long names I had given them to be able to clarify my uncertainties.

The configuration of the devices can be done via Web UI in the first boot phase of the gateway or does that phase only concern the configuration of the wifi and the mqtt broker?

If I can not do it via Web UI I have to:
  create a config.json file in the sketch data folder
  upload it to the SPIFFS file system using Tools> ESP8266 Sketch Data Upload?

Again I'm sorry if I'm stressful but I would like to avoid frying my LoLin NodeMcu!

Thanks

**Comment by Maddmax76** [comment]
according to the indications of

https://homieiot.github.io/homie-esp8266/docs/2.0.0/configuration/json-configuration-file/

the file should be like this:

{
  "name": "433toMQTT-Gateway",
  "device_id": "NodeMCU_1",
  "device_stats_interval": 60,
  "wifi": {
    "ssid": "MySSID",
    "password": "MyPWD"
  },
  "mqtt": {
    "host": "MyBroker",
    "port": 1883,
    "base_topic": "devices/",
    "auth": true,
    "username": "MyUser",
    "password": "MyPassword"
  },
  "ota": {
    "enabled": true
  },
  { "settings": { "channels": "Button-Engage:[139400];Button-Disengage:[139707];Button-Perimetric:[1398097];Button-Panic:[139803];Door-Entry":[524745];Door-Kitchen:[524745];“ } }
}
but all without spaces and on a line?

**Comment by mhaack** [comment]
The Home sample has the full config you must provide on the initial config. If you update config via MQTT you can also do partial updates and only send the settings.

**Comment by Maddmax76** [comment]
Good morning.
I entered the configuration via OTA, if I understood correctly what to put in the configuration, in the end it was quite simple....


But unfortunately:

- the gateway integrates Homie 2.0.1, hopenHAP has the binding for Homie 3.x, when I create the thing it marks it as OFFLINE because it lacks $ state, is it a typical parameter of version 3?
java.lang.Exception: Did not receive mandatory topic topic: homie / gw01 / $ state

- feedback on the NodeMCU the same problem that reflected on the sketch of 1 Technophile 433toMQTTto433: it seems that it does not receive data from the radiofrequency.
The receiver mounted on an ArduinoNano with the RCSwitch library works, in fact I was assuming to use the nano as a 433toSerial gateway.
Assuming that the NodeMcu was having problems I tried to change sketches, I found another code that uses RCSwitch to analyze the RF signals, and it works.
As you load your code or that of 1 Technophile the nodemcu part, connects to the wifi, connects to the broker mqtt, sends the keepalive signals and waits for signals, but receives nothing, even trying to change the pin where I connect the RF receiver


Do you have any idea?
![20190303_171335](https://user-images.githubusercontent.com/47970623/53719979-db165580-3e5f-11e9-8362-2aa59d8ca0e0.jpg)

**Comment by mhaack** [comment]
https://github.com/homieiot/homie-esp8266 does not support Homie V3 - yet. That is still in development. So auto discovery does not work.

**Comment by Maddmax76** [comment]
It's a pity, I'll wait impatiently for the new version then.
Thanks a lot!

**Comment by mhaack** [comment]
You don't have to. It works with OpenHab as well, only no discovery but you can configure it in OpenHab manually. 
Release of Homie impl. was not that often in the past, so I guess you will have to wait a while until that is implemented. Also this might need me to change the code of the mqtt-433mhz-gateway-homie to adapt. That said it will take time, I would not wait.

**Comment by Maddmax76** [comment]
humm, is there somewhere explained how to create Button-like objects on openHab that read my channels?

**Comment by Maddmax76** [comment]
apart from the fact that I must first make it work .... I do not know why but the version of RCSwitch that is in band with the gateway code does not work with my nodemcu, now I'm trying with the test sketch ReceiveDemo_Simple, but it does not go !

**Comment by mhaack** [comment]
You can ask that on https://community.openhab.org but that should be possible, at the end it is only about sending or receiving MQTT commands in OH.

**Comment by Maddmax76** [comment]
ok, thanks, Bye!

**Comment by Maddmax76** [comment]
hello, I solved the problem of reception, in practice the nodeMcu wants pin 13 (D7) to receive the RF. Done that works everything beautifully. The RF signals were read and transformed into MQTT messages that I read with HoenHAB.
The doubt is that in the configuration of the gateway I have set only one channel, the channel is used to create a mqtt topic where the values of the received RF signals are written.
Can I create more channels to get a topic for every RF device I use? And can I do it using the mqtt commands? Does it make sense or does not it suit?

Thanks bye!

**Comment by Maddmax76** [comment]
P.S. is there a way to download from the gateway the configuration file to be modified with a pc and then upload to the gateway with curl?

Thank you

**Comment by mhaack** [comment]
> Can I create more channels to get a topic for every RF device I use?

Yes same way like channels for buttons they work in both ways. 

> And can I do it using the mqtt commands? 

Yes, see Homie doc. But it is always updating the entire settings only.

**Comment by mhaack** [comment]
> P.S. is there a way to download from the gateway the configuration file to be modified with a pc and then upload to the gateway with curl?

You can do this using MQTT client. The gateway publishes the config during start. It is also logged to the console if you connect it via USB. A real download does not exist. Also CURL upload not only via MQTT.

**Comment by Maddmax76** [comment]
hello Markus, good morning.

I took a look at the configuration printed via the console:

```
💡 Firmware mqtt-gateway-livingroom (1.0.0)
🔌 Booting into normal mode 🔌
{} Stored configuration
  • Hardware device ID: a020a6251b5a
  • Device ID: gw01
  • Name: gwp1
  • Device Stats Interval: 60 sec
  • Wi-Fi: 
    ◦ SSID: not shown
    ◦ Password not shown
  • MQTT: 
    ◦ Host: 192.168.1.17
    ◦ Port: 1883
    ◦ Base topic: homie/
    ◦ Auth? yes
    ◦ Username: not shown
    ◦ Password not shown
  • OTA: 
    ◦ Enabled? yes
  • Custom settings: 
    ◦ channels: A:[407052];B:[86292];C:[12663];D:[253233];E:[426330];F:[145665];G:[524745];H1:[429246];H2:[429228];H3:[429222];H4:[429220];I1:[102249];I2:[102231];I3:[102225];I4:[102223];L1:[102249];L2:[102231];L4:[102223];M1:[311013];M2:[310968];M3:[310962];M4:[310960]; (set)
↕ Attempting to connect to Wi-Fi...

```
ho dato anche un'occhiata alla configurazione di Homie:

https://homieiot.github.io/homie-esp8266/docs/2.0.0/configuration/json-configuration-file/

but I think I'm looking at the wrong documentation.

I still do not understand how to define objects.

How can I indicate that A is a sensor, which must have its own topic and that when I receive a radio signal that reads 407052, homie must write in the PIPPO topic the value ON?

How to tell him instead that H is a remote control that sends 4 different radio stations: ON(429246), OFF(429228), PERIMETER(429222), PANIC(429220)?

I can not find any syntax to use or to report on how the logic of the program behaves.

Thanks, ciao!

**Comment by mhaack** [comment]
Hello,

the config looks ok. The channels work in both ways sender & receiver. If ran RF signal is received the gateway will transform that into MQTT message. On your remote control, if you press `ON` you should get some MQTT message on channel `H1` like `homie/gwp1/channel-H1` with the message send by the remote. 
Same applies for the sensor. If on `407052` a RF signal is received you get an MQTT message on `homie/gwp1/channel-A` with the sensor value.

Best is to test with some MQTT client like: `mosquitto_sub -h 192.168.1.17 -v -t '#'` - plus username and password and see which messages are received.

Markus

**Comment by Maddmax76** [comment]
Hi Markus, thanks for the information.

This morning I completed the configuration using the published information as the gateway publishes and working on the openHAB rules.

Now I managed to get what I wanted, a system parallel to my alarm system that can give me information on what triggered the alarm at home.

I only have to complete the gateway with the RF transmitter in order to activate or deactivate the alarm.

Thanks again for the suggestions and the patience and congratulations for the excellent work done with the gateway.
